docs/ADR/2019-11-27-firestore-saving-timed-transcripts.md

```diff
@@ -0,0 +1,227 @@
+* Bad, because paragraph breakdown subject to change
+* … <!-- numbers of pros and cons can vary -->
+
+### 3. Break down data in firestore - every word is a document
+
+An alternative could also be to  save each word as it's on document in a collection of words. Same thing could be done for paragraphs.
+
+However it could get very expensive to do a read of a whole document, as we'd need to fetch all the words etc.. and we'd get charge for every read. So even for a 20 minute transcript, with 3581 words, this could incur a substantial cost.
+
+* Good, because easy to reason around
+* Bad, because could get expensive, as it would increase number of reads
+
+
+### 4. Saving json in Firebase cloud storage?
+
+Another option is to use firestore triggers and/or [HTTPS callable functions](https://firebase.google.com/docs/functions/callable) to save the DPE json format into cloud storage.
+
+There would be no size limit, but we would not be able to take advantages of firebase functionalities for the data. 
+
+In the current client, when saving a timed transcript the whole payload is returned, there is currently no option to save intermediate results, or only what's the latest change. 
+
+This might change and get optimised in the future, but we also just want a solution that works for the now, and then look at ahead at further refactor and optimization.
+
+* Good, because file size limit on cloud storage is more then enough at [5TB](https://cloud.google.com/storage/quotas)
+* Bad, because introduces cost of calling a firebase cloud function to perform the operation
+* Bad, because it might introduce a delay (?)
+
+### 5. Compression - JSONB 
+Another option is to compress the json. eg with [js-bson](https://github.com/mongodb/js-bson)
+
+If we consider that 1 hour in DPE format is 1.3 MB, and for reference in GCP STT json format it is 2mb. After BSON compressed is 836 KB.
+
+As we so in the data structure example above for DPE json, paragraphs and words could be separated.
+
+So if you do words only then it's 1.1MB compressed to 822KB.
+And paragraphs that are 15KB in DPE json, and after JSONB compression they go down to 11KB.
+
+So I'd say this doesn't help much (?).
+
+Hard to see how this compression option would scale.
+
+* Good, because for around an hour worth of transcript, could get it below 1mb
+* Bad, because Hard to see how this compression option would scale.
+
+### 6. Compression - JSONC
+
+[github JSONC](https://github.com/tcorral/JSONC), [npm jsonc](https://www.npmjs.com/package/jsonc)
+
+> Be careful with this method because it's really impressive if you use it with a JSON with a big amount of data, but it could be awful if you use it to compress JSON objects with small amount of data because it could increase the final size.
+
+article from [coderwall, compress your json data](https://coderwall.com/p/mekopw/jsonc-compress-your-json-data-up-to-80)
+
+```js
+const fs = require('fs')
+const jsonc = require('jsonc');
```

Comment:
https://www.npmjs.com/package/jsoncomp <- the actual library

I'll get this script to work and let you know.